internal/command/cliconfig/ociauthconfig/docker_cli_credentials_config.go

```diff
@@ -0,0 +1,291 @@
+	if authsPropertyName == "" {
+		return NoCredentialsSpecificity // Invalid
+	}
+	var gotDomain, gotRepositoryPath string
+	if strings.Count(authsPropertyName, "/") > 0 {
+		var err error
+		gotDomain, gotRepositoryPath, err = ParseRepositoryAddressPrefix(authsPropertyName)
+		if err != nil {
+			// Since these Docker CLI-style config files are not exclusively for OpenTofu,
+			// we just silently ignore a property whose name we can't make sense of
+			// in case a future version of this format adds something new that we
+			// don't yet support.
+			return NoCredentialsSpecificity // Invalid
+		}
+	} else {
+		gotDomain = authsPropertyName
+	}
+	if gotDomain != wantRegistryDomain {
+		return NoCredentialsSpecificity // does not match
+	}
+	if gotRepositoryPath == "" {
+		return DomainCredentialsSpecificity // matches only the domain
+	}
+	// If authsPropertyName includes a path (that is: if gotRepositoryPath != "")
+	// then we need to test if gotRepositoryPath has at most as many segments
+	// as wantRepositoryPath, and that those segments all match.
+	wantSegCount := strings.Count(wantRepositoryPath, "/") + 1
+	gotSegCount := strings.Count(gotRepositoryPath, "/") + 1
+	if gotSegCount > wantSegCount {
+		// The path in authsPropertyName has too many segments to possibly
+		// match wantRepositoryPath.
+		return NoCredentialsSpecificity
+	}
+	wantRemain := wantRepositoryPath
+	gotRemain := gotRepositoryPath
+	for range wantSegCount {
+		var thisWant, thisGot string
+		var moreGot bool
+		thisWant, wantRemain, _ = strings.Cut(wantRemain, "/")
+		thisGot, gotRemain, moreGot = strings.Cut(gotRemain, "/")
+		if thisWant != thisGot {
+			// If we find a mismatch before we run out of "got" then
+			// only a prefix of authsPropertyName matches the wanted
+			// address, which means that this auth configuration is
+			// ineligible for the given repository.
+			return NoCredentialsSpecificity
+		}
+		if !moreGot {
+			break // no more segments to compare, so we've found at least a valid prefix match
+		}
+	}
+	// If we get here without returning early then the whole of gotRepositoryPath
+	// matched, so our specificity is the number of segments in that path.
+	return RepositoryCredentialsSpecificity(uint(gotSegCount)) //nolint:gosec // number of path segments in the input cannot possibly be negative
+}
+
+func dockerCLIStyleAuthFileSearchLocations(env ConfigDiscoveryEnvironment) iter.Seq[string] {
+	// The following enumerates all of the search locations described in the following
+	// document, taking into account the environment's reported operating system name:
+	// https://github.com/containers/image/blob/main/docs/containers-auth.json.5.md
+	os := env.OperatingSystemName()
+	homeDir := env.UserHomeDirPath()
+	isMacOS := os == "darwin"
+	isWindows := os == "windows"
+	isLinux := os == "linux"
+	return func(yield func(string) bool) {
```

Comment:
Nit: Does the complexity of an iter.Seq benefit us here?  We aren't memory or CPU bound by this function.